### PR TITLE
[SPARK-37650][PYTHON] Tell spark-env.sh the python interpreter

### PR DESCRIPTION
**Pull request by PerilousApricot**
When loading config defaults via spark-env.sh, it can be useful to know
the current pyspark python interpreter to allow the configuration to set
values properly. Pass this value in the environment as
_PYSPARK_DRIVER_SYS_EXECUTABLE to the environment script.

### What changes were proposed in this pull request?

It's currently possible to set sensible site-wide spark configuration defaults by using `$SPARK_CONF_DIR/spark-env.sh`. In the case where a user is using pyspark, however, there are a number of things that aren't discoverable by that script, due to the way that it's called. There is a chain of calls (java_gateway.py -> shell script -> java -> shell script) that ends up obliterating any bit of the python context.

This change proposes to add en environment variable `_PYSPARK_DRIVER_SYS_EXECUTABLE` which points to the filename of the top-level python executable within pyspark's `java_gateway.py` bootstrapping process. With that, spark-env.sh will be able to infer enough information about the python environment to set the appropriate configuration variables.

### Why are the changes needed?

Right now, there a number of config options useful to pyspark that can't be reliably set by `spark-env.sh` because it is unaware of the python context that spawning the executor. To give the most trivial example, it is currently possible to set `spark.kubernetes.container.image` or `spark.driver.host` based on information readily available from the environment (e.g. the k8s downward API). However, `spark.pyspark.python` and family cannot be set because when `spark-env.sh` executes it's lost all of the python context. We can instruct users to add the appropriate config variables, but this form of cargo-culting is error-prone and not scalable. It would be much better to expose important python variables so that pyspark can not be a second-class citizen.

### Does this PR introduce _any_ user-facing change?

Yes. With this change, if python spawns the JVM, `spark-env.sh` will receive an environment variable `_PYSPARK_DRIVER_SYS_EXECUTABLE` pointing to the python executor.

### How was this patch tested?

To be perfectly honest, I don't know where this fits into the testing infrastructure. I monkey-patched a binary 3.2.0 install to add the lines to java_gateway.py and that works, but in terms of adding this to the CI ... I'm at a loss. I'm more than willing to add the additional info, if needed.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
Mind filing a JIRA please? see also https://spark.apache.org/contributing.html

**Comment by PerilousApricot** [comment]
@HyukjinKwon I was already trying to figure out the JIRA :) I'm not as good at that as I am with GH

**Comment by PerilousApricot** [comment]
@HyukjinKwon https://issues.apache.org/jira/browse/SPARK-37650 thanks!

**Comment by HyukjinKwon** [comment]
Thanks @PerilousApricot. Let's link the JIRA into the PR title:

> Open a pull request against the master branch of apache/spark. (Only in special cases would the PR be opened against other branches.)
The PR title should be of the form [SPARK-xxxx][COMPONENT] Title, ...

**Comment by PerilousApricot** [comment]
@HyukjinKwon I fixed it! Sorry for being out of order and making the PR before the JIRA

**Comment by PerilousApricot** [comment]
@HyukjinKwon dunno if this is the right person to ask, but since you're the pyspark guy, are you amenable to having a `spark-env.py` that gets loaded from the pyspark context before the java gateway gets started? PySpark is at a disadvantage compared with other languages because the config hooks happen so late in the process. It would be nice to set configs in java_gateway.py before the JVM is started, compared with now where pyspark initialization happens very late

File: docs/configuration.md [hunk 1]
```diff
@@ -3029,7 +3029,10 @@ can be found on the pages for each mode:
 Certain Spark settings can be configured through environment variables, which are read from the
 `conf/spark-env.sh` script in the directory where Spark is installed (or `conf/spark-env.cmd` on
 Windows). In Standalone and Mesos modes, this file can give machine specific information such as
-hostnames. It is also sourced when running local Spark applications or submission scripts.
+hostnames. It is also sourced when running local Spark applications or submission scripts. For
+pyspark applications, the environment variable `_PYSPARK_DRIVER_SYS_EXECUTABLE` will be set to
+the python interpreter's `sys.executable`, which will allow further customization based on the
+user's virtual environment.
```
Comment (itholic) [inline docs/configuration.md:3035]: nit: Can we use official naming for Python and PySpark ??

- pyspark -> PySpark
- python -> Python

Comment (PerilousApricot) [reply docs/configuration.md]: done @ e747a2f

Comment (PerilousApricot) [reply docs/configuration.md]: done!

**Comment by PerilousApricot** [comment]
Hello @HyukjinKwon and Happy New year! Now that I'm back to work after the holiday, I was wondering if you had any more thoughts on this PR.

**Comment by PerilousApricot** [comment]
Hi @HyukjinKwon do you have any other thoughts on this PR? It would be nice to get it into a release at some point

**Comment by MaxGekk** [comment]
@HyukjinKwon @itholic Should we postpone Spark 3.3 because of this PR?

**Comment by itholic** [comment]
> Yes. With this change, if python spawns the JVM, spark-env.sh will receive an environment variable _PYSPARK_DRIVER_SYS_EXECUTABLE pointing to the python executor.

Would you ming elaborating that what is the different from `PYSPARK_PYTHON` or `PYSPARK_DRIVER_PYTHON ` ??

Ain't they already pointing to the python executor ??

**Comment by PerilousApricot** [comment]
Those are spark configuration variables, but importantly they are not passed to spark-env.sh, so you can't use them to bootstrap python -- by then it's too late. This change exposes the python interpreter to the environment customization script so the right thing can be done early

**Comment by itholic** [comment]
@PerilousApricot thanks!

Then the changes themselves look fine to me. But I'm not sure how can we do the test since I don't have enough background here.

Can you provide a simple code example where this change would be really useful?

**Comment by PerilousApricot** [comment]
Hi @itholic -- the following trivial example should work for conf/spark-env.sh

```
export PYSPARK_PYTHON="${_PYSPARK_DRIVER_SYS_EXECUTABLE}"
```

**Comment by PerilousApricot** [comment]
I'm not sure of what else needs to be done. I made the PR last year and
have asked repeatedly for it to be looked at. It's a small PR, but I can't
get anyone to look at it.--
It's dark in this basement.

**Pull request closed without merging**